### ISSUE TITLE
New Resource: okta_app_push_groups

### ISSUE THREAD

**Issue by ktham** [okta/terraform-provider-okta#1409]
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Currently, each time a new Okta group is created through Terraform, an engineer is required to _manually_ go into the Okta Web UI to configure a push group for each Okta application in order to push the new Okta group to those SSO applications.

This is undesired overhead. We would like to follow the infrastructure-as-code model for managing _all_ Okta configurations. We would like to see this Okta provider implement a resource to programmatically configure Push Groups.

### New Resource(s)

- okta_app_push_groups

### Potential Terraform Configuration

```hcl
resource "okta_app_push_groups" "aws_sso_active_push_groups" {
  app_id = data.okta_app.aws_sso.id

  group_names = [
    okta_group.admin.name,
  ]

  push_immediately = true
}

data "okta_app" "aws_sso" {
  label = "AWS SSO"
}

resource "okta_group" "aws_admin" {
  name        = "aws-admin"
  description = "AWS Admin Group (Managed by Terraform)"
}

```

**Comment by noinarisak** [comment]
Hey @ktham, to my knowledge and thanks to @gabrielsroka, that particular object, `grouppush` is an internal endpoint.

Unfountantely, which @monde could add more details, the process/procedure is to only have Official endpoints that are published on [dev.okta.com/ref](https://developer.okta.com/docs/reference/) which would outline the details about the endpoint and json payloads and structure.

I think the next course of action is to submit https://ideas.okta.com/ so your request is visibility to our Engineering Team.

**Comment by ktham** [comment]
Hi @noinarisak , thanks for the quick follow up!

I see, given the current constraints, are there any workarounds available to get the groups created by Terraform to make it to the Okta application(s) without requiring manual steps through a Web GUI?

And @monde , to confirm, there is an API that _could_ be used to implement this Terraform resource, but it is not an _official_ endpoint? (And the current policy is to only expose it via the web GUI rather than your okta terraform provider?)

> I think the next course of action is to submit https://ideas.okta.com/ so your request is visibility to our Engineering Team.

Before I create a new request, is there already an existing feature request like this that I can follow and +1 on? (Is it this one? https://ideas.okta.com/app/#/case/122821) I would have assumed other folks using the Terraform provider would have run into the same problem, I'm surprised there hasn't been more discussion on this on the TF provider side

Thanks again for the follow up @noinarisak !

**Comment by gabrielsroka** [comment]
in the browser, i opened Dev Tools > Network, then i created an AD Push Group using the Okta Admn Console to see what it did. 

`0oa...` below is the app ID:

```
POST /api/internal/instance/0oa.../grouppush

{
    "status": "ACTIVE",
    "userGroupId": "00g...",
    "newAppGroupName": "ad push group",
    "groupPushAttributes": {
        "groupScope": "GLOBAL",
        "groupType": "SECURITY",
        "distinguishedName": "ou=ou,ou=adusers,dc=gsroka,dc=local",
        "samAccountName": "ad push group"
    }
}
```

**Comment by monde** [comment]
Right, we can't hit internal endpoints `/api/internal`, sadly.

**Comment by monde** [comment]
Okta internal reference: https://oktainc.atlassian.net/browse/OKTA-563717

**Comment by brian-nomi** [comment]
This has been an open issue since a 2021 bug. While we can all appreciate 'it's not on the roadmap, pls submit an idea' - it's been two years. Okta is aware. Fix it.

**Comment by ChefAustin** [comment]
Doesn't sound like this PR will get merged given that it leverages private/internal API's but, for whatever its worth, this _would_ fix #312.

Also noteworthy is that in addition to Okta Idea's submission [#122821](https://ideas.okta.com/app/#/case/122821), there is also [#167182](https://ideas.okta.com/app/#/case/167182) that I submitted 3 weeks ago requesting this exact thing.

**Comment by ktham** [comment]
Thank you for following up @monde! It's great that it's being tracked in an internal JIRA ticket, but any hints as to whether this is indeed being reviewed or discussed internally? It seems that a good number of folks have requested this or something related for push groups for ~ 3 years now, but it doesn't seem to be getting any traction and wondering if this is indeed being looked at at all

**Comment by hi-artem** [comment]
This really MUST be an external API. Please Okta fix this.

**Comment by ragona** [comment]
This is a significant gap in the provider, you really cannot get core functionality working as IaC right now because you need to go click buttons in the UI. Please consider prioritizing this one.

**Comment by joshgch** [comment]
This feature is also needed within my organization.

**Comment by haorong4** [comment]
This feature is really needed by my project. Okta should prioritize this.

**Comment by Alex-Rockliff** [comment]
This is also a really important feature for my organisation.

**Comment by ChefAustin** [comment]
I'm going to knowingly go against the guidance in this Issue's initial "Community Note" comment regarding 👍 to simply reiterate to folks what is said there:

**Please stop responding to this thread with "+1", "my org too..." comments; all you need to do to increase impact is merely smash that 👍 emoji reaction button on the initial Issue comment. These additional comments don't do anything to increase impact/priority and -- if we, the Okta user community, want to see this actually implemented -- we should be doing our best to be good rule-abiding citizens.**

Conversely, if you really want to attempt to raise priority of this request, I would suggest reaching out to your organization's account representative and push them (that said, I am not an Okta employee and this might be bad advice... but it sure is something that I've been doing).

**Comment by abhinav-khanna-ib** [comment]
We have rules to push Okta groups to AWS. In our case, we were seeing similar behavior that group was created but not pushed to AWS unless someone goes in and click on the group and just saves it.

It was access-related. We granted app admin to the terraform user and it worked.

**Comment by ChefAustin** [comment]
Has this request been given any more consideration from the Okta side?

**Comment by MrYeske-SA** [comment]
Checking on this as well. I've been asking my Okta Partner rep about Programmatic Okta Application Provisioning Configuration (API Integration / Group Push / Group Push Rules Configurations) for YEARS. What do we need to get it supported on the API side, so we can then talk about the merits of it being added to the Terraform provider. Almost everything related to this over the years has been Closed.

**Comment by bbreckenridge** [comment]
My team is also trying to automate the entire setup and configuration for new Okta Applications.  Why doesn't the API support Okta Push Groups / Rules?

**Comment by markekibbe** [comment]
My group as well has been trying to automate this, it would prove helpful if it was implemented.

**Comment by monde** [comment]
CC'ing our PM @jefftaylor-okta to speak on the status of this feature upstream in the Okta Management API

**Comment by MrYeske-SA** [comment]
> CC'ing our PM @jefftaylor-okta to speak on the status of this feature upstream in the Okta Management API

Ready and waiting...

**Comment by swade1987** [comment]
This needs to be an external API its causing issues at the company I work at as well, we try to do everything is code and can't in this instance 😢

**Comment by ktham** [comment]
Hi @monde / @jefftaylor-okta ,

Is Okta still considering our request? I noticed that in https://ideas.okta.com/app/#/case/122821, that it was updated  and your team is now rejecting this feature request for programmatically managing Okta Push Groups.

<img width="449" alt="Screenshot 2024-02-27 at 1 31 48 PM" src="https://github.com/okta/terraform-provider-okta/assets/882701/ed4dab0f-f0ea-4d8e-8985-7059ad847ec1">

If so, that would be very disappointing, can someone on your team confirm the status?

**Comment by iniinikoski** [comment]
It's been a year now since the last update here. Unbelievable that this is still open. Okta should get their stuff together. Will definitely raise this to our account manager.

**Comment by dgruber-nydig** [comment]
> A public Push Groups API is scheduled to be released in Calendar Q3 2025.

🎊

**Comment by ktham** [comment]
Indeed, this is good news! Attached is the current screenshot from https://ideas.okta.com/app/#/case/122821 

<img width="618" alt="Image" src="https://github.com/user-attachments/assets/359f95db-1bf7-4efd-9076-e9689a58b6db" />

**Comment by DustinMoriarty** [comment]
+1 for this. We are considering moving away from Okta for SCIM because of this gap. I am glad it is on Q3 2025. However, it may be too little too late for us. 

We take the lack of a public API for configuring SCIM as a signal that Okta does not consider SCIM to be part of it's core business products so we are concerned about overall support for SCIM.

**Comment by wrprice** [comment]
Push group API is now public preview, with GA expected in September 2025 per an update on Okta Ideas

https://developer.okta.com/docs/api/openapi/okta-management/management/tag/GroupPushMapping/

**Comment by zepeng811** [comment]
Perfect timing and thanks for the update! I just found this thread today as my org also have similar issues, good to hear it's in public preview and will GA soon.

**Comment by aditya-okta** [comment]
We will be releasing new version of the terraform provider soon where we can manage Group Push Mappings . Thanks.

**Comment by aru-trackunit** [comment]
> We will be releasing new version of the terraform provider soon where we can manage Group Push Mappings . Thanks.

I see that the ticket is still open and when browsing https://registry.terraform.io/providers/okta/okta/latest I can't see `okta_app_push_groups` resource, am I missing something? Thanks in advance!

**Comment by aditya-okta** [comment]
Hey @aru-trackunit , we still haven't released it yet. We are currently working on adding a bunch of more resources and data sources, so we'll be adding it soon.Thanks.

**Comment by aron-muon** [comment]
This PR is the one to track - it adds the group push API
https://github.com/okta/okta-sdk-golang/pull/538

**Comment by aron-muon** [comment]
Now that the golang v6 SDK was added by https://github.com/okta/terraform-provider-okta/pull/2553 this feature request should be unblocked within this repo.

**Comment by pranav-okta** [comment]
@aron-muon Yes, we are working on adding new resources and updating the SDK in the Terraform provider.

**Comment by tim-koehler** [comment]
@pranav-okta The APIs are now available and v6 SDK is present in the provider. This means that there are no blockers for this feature anymore.
Are you at Okta already working on this?

Otherwise I can find some time to implement the resources to stop all of us suffering.

**Comment by tim-koehler** [comment]
@pranav-okta  @aditya-okta 
Please have a look at my PR regarding this. Looking forward to your feedback🙂
